drivers/dac/dac_dacx0501.c

```diff
@@ -0,0 +1,197 @@
+ * bits resolution respectively. Data sheet can be found here:
+ * https://www.ti.com/lit/ds/symlink/dac80501.pdf
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/dac.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/byteorder.h>
+
+LOG_MODULE_REGISTER(dac_dacx0501, CONFIG_DAC_LOG_LEVEL);
+
+#define DACX0501_REG_DEVICE_ID 0x01U
+#define DACX0501_REG_SYNC      0x02U
+#define DACX0501_REG_CONFIG    0x03U
+#define DACX0501_REG_GAIN      0x04U
+#define DACX0501_REG_TRIGGER   0x05U
+#define DACX0501_REG_STATUS    0x07U
+#define DACX0501_REG_DAC       0x08U
+
+#define DACX0501_MASK_DEVICE_ID_RES      GENMASK(14, 12)
+#define DACX0501_MASK_CONFIG_REF_PWDWN   BIT(8)
+#define DACX0501_MASK_CONFIG_DAC_PWDWN   BIT(0)
+#define DACX0501_MASK_GAIN_BUFF_GAIN     BIT(0)
+#define DACX0501_MASK_GAIN_REFDIV_EN     BIT(8)
+#define DACX0501_MASK_TRIGGER_SOFT_RESET (BIT(1) | BIT(3))
+#define DACX0501_MASK_STATUS_REF_ALM     BIT(0)
+
+/* Specifies the source of the reference voltage. */
+enum ref_voltage_source {
+	REF_INTERNAL, /* Internal 2.5V voltage reference. */
+	REF_EXTERNAL, /* External pin voltage reference. */
+};
+
+/* Specifies the reference voltage multiplier. */
+enum voltage_mult {
+	VM_MUL2, /* Multiplies by 2. */
+	VM_MUL1, /* Multiplies by 1. */
+	VM_DIV2, /* Multiplies by 0.5 */
+};
+
+struct dacx0501_config {
+	struct i2c_dt_spec i2c_spec;
+	enum ref_voltage_source ref_source;
+	enum voltage_mult voltage_mult;
+};
+
+struct dacx0501_data {
+	/* Device can only be configured once. */
+	bool configured;
+	/* Number of bits in the DAC register: Either 12, 14 or 16. */
+	uint8_t resolution;
+};
+
+static int dacx0501_reg_read(const struct device *dev, const uint8_t addr, uint16_t *data)
+{
+	const struct dacx0501_config *config = dev->config;
+	uint8_t raw_data[2];
+	int status;
+
+	status = i2c_write_read_dt(&config->i2c_spec, &addr, sizeof(addr), raw_data,
+				   sizeof(raw_data));
+	if (status) {
+		return status;
+	}
+
+	/* DAC is big endian. */
+	*data = sys_get_be16(raw_data);
+	return 0;
+}
+
+static int dacx0501_reg_write(const struct device *dev, uint8_t addr, uint16_t data)
+{
+	const struct dacx0501_config *config = dev->config;
+	uint8_t write_cmd[3] = {addr};
+
+	/* DAC is big endian. */
+	sys_put_be16(data, write_cmd + 1);
+
+	return i2c_write_dt(&config->i2c_spec, write_cmd, sizeof(write_cmd));
+}
+
+static int dacx0501_channel_setup(const struct device *dev,
+				  const struct dac_channel_cfg *channel_cfg)
+{
+	const struct dacx0501_config *config = dev->config;
+	struct dacx0501_data *data = dev->data;
+	uint16_t device_id;
+	int status;
+
+	/* DACx0501 series only has a single output channel. */
+	if (channel_cfg->channel_id != 0) {
+		LOG_ERR("Unsupported channel %d", channel_cfg->channel_id);
+		return -ENOTSUP;
+	}
+
+	if (data->configured) {
+		LOG_ERR("Channel already configured");
+		return -EINVAL;
```

Comment:
This return code is not exactly right considering the API docs. As mentioned above, unless the device can really only be configured once, I don't think this is an error.

Reply:
Done